### PR TITLE
Automated Testing

### PR DESCRIPTION
**Pull request by JuanGonzalezCaminero**
* Python script to run tests defined in JSON format
* Several python scripts for plotting test results
* Result validation and benchmarking options for the tests
* Extended the random gun functionality with the ability to specify per-particle energy

In order to run an example benchmark or validation, compile AdePT with ```TEST=ON```, and run the following in the ```testing``` directory:

```python3 run_test.py benchmark_config.json```
```python3 run_test.py validation_config.json```

It will run some simple example tests and generate plots from the results.

### CODE REVIEW

**Comment by agheata** [comment]
Please provide a description for the PR, including the features introduced, how to run, etc. Ideally, you can put this info in also in a README file for the benchmark directory

File: benchmark/benchmark_config.json [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+                        "randomize_gun" : true,
+                        "num_events" : 5
+                    },
+                    
+                    "random_gun_configuration" :
+                    {
+                        "particles" :
+                        {
+                            "e-" : {"weight" : 0.4, "energy": "15 GeV"},
+                            "e+" : {"weight" : 0.2, "energy": "5"},
```
Comment (agheata) [inline benchmark/benchmark_config.json:56]: Units should be mandatory for energy

File: examples/Example17/macros/example17.mac.in [hunk 2]
```diff
@@ -136,8 +136,9 @@
 /example17/gun/print true
 
 #If false, the following parameters are ignored
-/example17/gun/randomizeGun false
-/example17/gun/addParticle e-
+/example17/gun/randomizeGun true
+#Usage: /example17/gun/addParticle [type] [weight] [energy]
+/example17/gun/addParticle e- -1 10 GeV
```
Comment (agheata) [inline examples/Example17/macros/example17.mac.in:141]: Not obvious what `-1` means

File: examples/Example17/macros/example17.mac.in [hunk 3]
```diff
@@ -136,8 +136,9 @@
...
 /example17/gun/addParticle e+
```
Comment (agheata) [inline examples/Example17/macros/example17.mac.in:142]: If weight is missing, I supposed you redistribute equally the remaining weight as we discussed, but if the energy is missing what is happening?

File: examples/Example17/src/PrimaryGeneratorMessenger.cc [hunk 4]
```diff
@@ -55,8 +56,13 @@ PrimaryGeneratorMessenger::PrimaryGeneratorMessenger(PrimaryGeneratorAction *Gun
...
   fAddParticleCmd = new G4UIcmdWithAString("/example17/gun/addParticle", this);
-  fAddParticleCmd->SetGuidance("When using randomization, add a particle to the list of possibilities, with the option "
-                               "to provide a weight to indicate how often it should appear");
+  fAddParticleCmd->SetGuidance("When using randomization, add a particle to the list of possibilities\n\
+                                Usage: /example17/gun/addParticle [type] [weight] [energy] [energy unit]\n\
+                                type: particle name\n\
+                                weight: probability that the particle will appear, between 0 and 1\n\
+                                energy: energy for this type of particle\n\
+                                Note: The default unit applies to the energy parameter.\n\
+                                Note: In order to specify an energy but no weight, use a weight of -1\n");
```
Comment (agheata) [inline examples/Example17/src/PrimaryGeneratorMessenger.cc:65]: OK, I see what -1 means. To make it nicer, during the argument tokenization you could enforce pairs key+value, so if a key like `weight` is missing you would not need to put -1.

**Review by agheata — CHANGES_REQUESTED** [review]
Just a first set of comments

File: examples/Example17/macros/example17.mac.in (hunk 3 above)
Comment (JuanGonzalezCaminero) [reply examples/Example17/macros/example17.mac.in:142]: It gets the energy set for the gun

File: examples/Example17/src/PrimaryGeneratorMessenger.cc (hunk 4 above)
Comment (JuanGonzalezCaminero) [reply examples/Example17/src/PrimaryGeneratorMessenger.cc:65]: I can change it to key-value yes, it should be more intuitive than now

File: examples/Example17/macros/example17.mac.in (hunk 3 above)
Comment (JuanGonzalezCaminero) [reply examples/Example17/macros/example17.mac.in]: Updated, now it's mandatory to provide the energy

**Review by agheata — APPROVED** [review]

**Review by agheata — CHANGES_REQUESTED** [review]
Leaving it unapproved until the pipeline passes

**Review by agheata — APPROVED** [review]

**Pull request merged**